service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,463 @@ import (
+		return nil, serviceerror.NewWorkflowExecutionAlreadyStarted(
+			"First workflow task is no longer inflight, retried request came in too late",
+			request.RequestId,
+			runID,
 		)
-		switch err {
-		case nil:
-			return &historyservice.StartWorkflowExecutionResponse{
-				RunId: runID.String(),
-			}, nil
-		case consts.ErrWorkflowCompleted:
-			// previous workflow already closed
-			// fallthough to the logic for only creating the new workflow below
-		default:
+	}
+
+	events, err := s.getWorkflowHistory(ctx, mutableStateInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.generateResponse(runID, mutableStateInfo.workflowTaskInfo, events)
+}
+
+// getMutableStateInfo gets the relevant mutable state information while getting the state for the given run from the
+// workflow cache and managing the cache lease.
+func (s *Starter) getMutableStateInfo(ctx context.Context, runID string) (*mutableStateInfo, error) {
+	// We techincally never want to create a new execution but in practice this should not happen.
+	workflowContext, releaseFn, err := s.workflowConsistencyChecker.GetWorkflowCache().GetOrCreateWorkflowExecution(
+		ctx,
+		s.namespace.ID(),
+		common.WorkflowExecution{WorkflowId: s.request.StartRequest.WorkflowId, RunId: runID},
+		workflow.CallerTypeAPI,
+	)
+	var releaseErr error
+	defer func() {
+		releaseFn(releaseErr)
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+
+	var mutableState workflow.MutableState
+	mutableState, releaseErr = workflowContext.LoadMutableState(ctx)
+	return extractMutableStateInfo(mutableState)
+}
+
+// extractMutableStateInfo extracts the relevant information to generate a start response with an eager workflow task.
+func extractMutableStateInfo(mutableState workflow.MutableState) (*mutableStateInfo, error) {
+	branchToken, err := mutableState.GetCurrentBranchToken()
+	if err != nil {
+		return nil, err
+	}
+
+	// Future work for the request retry path: extend the task timeout (by failing / timing out the current task).
+	workflowTaskInfo, hasInflight := mutableState.GetInFlightWorkflowTask()
+
+	return &mutableStateInfo{
+		branchToken:      branchToken,
+		lastEventID:      mutableState.GetNextEventID() - 1,
+		workflowTaskInfo: workflowTaskInfo,
+		hasInflight:      hasInflight,
+	}, nil
+}
+
+// getWorkflowHistory loads the workflow history based on given mutable state information from the DB.
+func (s *Starter) getWorkflowHistory(ctx context.Context, mutableState *mutableStateInfo) ([]*history.HistoryEvent, error) {
+	var events []*history.HistoryEvent
+	// Future optimization: generate the task from mutable state to save the extra DB read.
+	// NOTE: While unlikely that there'll be more than one page, it's safer to make less assumptions
+	for {
+		response, err := s.shardCtx.GetExecutionManager().ReadHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
```

Comment:
frontend also support return raw history event and it's controlled by a feature flag I think. Probably add a TODO for that here.

Reply:
Added todo above.